Linux/sgx/test_app/enclave/tests/evp_smx.c

```diff
@@ -36,591 +36,666 @@
+			ret = -2;
+			break;
+		}
+		// 3. Set the new EC key's curve
+		if (EC_KEY_set_group(ec_key, ec_group) != 1) {
+			printf("Error: fail to set the new EC key's curve\n");
+			ret = -3;
+			break;
+		}
+		// 4. Generate key pair based on the curve
+		if (!EC_KEY_generate_key(ec_key)) {
+			printf("Error: fail to generate key pair based on the curve\n");
+			ret = -4;
+			break;
+		}
+#elif OPENSSL_VERSION_NUMBER >= 0x30000000
+		printf("OpenSSL version: 3.0.0 or higher\n");
+		// 1. Create a EVP_PKEY_CTX for SM2
+		evp_pkey_ctx = EVP_PKEY_CTX_new_from_name(NULL, "SM2", NULL);
+		if (evp_pkey_ctx == NULL) {
+			printf("Error: fail to create a EVP_PKEY_CTX for SM2\n");
+			ret = -5;
+			break;
+		}
+		// 2. Initialize the SM2 EVP_PKEY_CTX
+		if (EVP_PKEY_keygen_init(evp_pkey_ctx) != 1) {
+			printf("Error: fail to initialize the SM2 EVP_PKEY_CTX\n");
+			ret = -6;
+			break;
+		}
+		// 3. Set params for the SM2 EVP_PKEY_CTX
+		OSSL_PARAM params[2];
+		params[0] = OSSL_PARAM_construct_utf8_string(OSSL_PKEY_PARAM_GROUP_NAME, "SM2", 0);
+		params[1] = OSSL_PARAM_construct_end();
+		if (EVP_PKEY_CTX_set_params(evp_pkey_ctx, params) != 1) {
+			printf("Error: fail to set params for SM2 EVP_PKEY_CTX\n");
+			ret = -7;
+			break;
+		}
+		// 4. Generate a SM2 EVP_PKEY by the SM2 EVP_PKEY_CTX
+		if (EVP_PKEY_generate(evp_pkey_ctx, &evp_pkey) != 1) {
+			printf("Error: fail to generate the SM2 EVP_PKEY by SM2 EVP_PKEY_CTX\n");
+			ret = -8;
+			break;
+		}
+#endif
+
+		// 5. Generate SM2 private key
+		pri_bio = BIO_new(BIO_s_mem());
+		if (pri_bio == NULL) {
+			printf("Error: fail to create a BIO for SM2 private key\n");
+			ret = -9;
+			break;
+		}
+#if OPENSSL_VERSION_NUMBER < 0x30000000
+		if (!PEM_write_bio_ECPrivateKey(pri_bio, ec_key, NULL, NULL, 0, NULL, NULL)) {
+			printf("Error: fail to write SM2 private key from ec_key to the BIO\n");
+			ret = -10;
+			break;
+		}
+#elif OPENSSL_VERSION_NUMBER >= 0x30000000
+		if (!PEM_write_bio_PrivateKey(pri_bio, evp_pkey, NULL, NULL, 0, NULL, NULL)) {
+			printf("Error: fail to write SM2 private key from evp_pkey to the BIO\n");
+			ret = -11;
+			break;
+		}
+#endif
+		pri_len = BIO_pending(pri_bio);
+		if (pri_len == 0) {
+			printf("Error: fail to get size of the BIO for SM2 private key\n");
+			ret = -12;
+			break;
+		}
+		*private_key = (char*)malloc(pri_len+1);
+		if (BIO_read(pri_bio, *private_key, pri_len) <= 0) {
+			printf("Error: fail to read SM2 private key from the BIO\n");
+			ret = -13;
+			break;
+		}
+		(*private_key)[pri_len] = '\0';
+
+		// 6. Generate SM2 public key
+		pub_bio = BIO_new(BIO_s_mem());
+		if (pub_bio == NULL) {
+			printf("Error: fail to create a BIO for SM2 public key\n");
+			ret = -14;
+			break;
+		}
+#if OPENSSL_VERSION_NUMBER < 0x30000000
+		if (!PEM_write_bio_EC_PUBKEY(pub_bio, ec_key)) {
+			printf("Error: fail to write SM2 public key from ec_key to the BIO\n");
+			ret = -15;
+			break;
+		}
+#elif OPENSSL_VERSION_NUMBER >= 0x30000000
```

Comment:
You can simply use only `#else` here, since it should be enough when it is `#if OPENSSL_VERSION_NUMBER < 0x30000000` above.